cpp/ql/src/semmle/code/cpp/ir/implementation/internal/TInstruction.qll

```diff
@@ -0,0 +1,102 @@
+private import TInstructionInternal
+private import IRFunctionBase
+private import TInstructionImports as Imports
+private import Imports::IRType
+private import Imports::Opcode
+
+/**
+ * An IR instruction. `TInstruction` is shared across all phases of the IR. There are individual
+ * branches of this type for instructions created directly from the AST (`TRawInstruction`) and for
+ * instructions added by each stage of SSA construction (`T*PhiInstruction`, `T*ChiInstruction`,
+ * `T*UnreachedInstruction`). Each stage then defines a `TStageInstruction` type that is a union of
+ * all of the branches that can appear in that particular stage. The public `Instruction` class for
+ * each phase extends the `TStageInstruction` type for that stage.
+ */
+cached
+newtype TInstruction =
+  TRawInstruction(
+    IRFunctionBase irFunc, Opcode opcode, Language::AST ast,
+    IRConstruction::Raw::InstructionTag1 tag1, IRConstruction::Raw::InstructionTag2 tag2
```

Comment:
Fixed. The RA diffs look mostly benign. I think that since we're caching the individual accessor predicates anyway, there's no real disadvantage to having outside of the IPA constructor.

Reply:
It turns out that _removing_ these from the IPA type also caused some consistency problems. Rather than fix them inline, I've opened a [separate PR](https://github.com/github/codeql/pull/3719) that should be merged first.

Reply:
I've merged #3719.